Fix numba NumbaTypeSafetyWarning and raise warnings as errors

### Version checks

- [x] I have checked that this issue has not already been reported.
- [x] I have checked that this bug exists on the latest version.

### Reproducible Example

The `tests/test_hydromt.py::test_model_build[piave]` and `tests/test_dflowfm.py::test_setup_rivers_from_dem` tests raise a NumbaTypeSafetyWarning because an unsafe cast is performed (unsafe cast from int32 to undefined). This happens in pyflwdir:
```python
def from_dataframe(df: "pandas.DataFrame", ds_col="idx_ds") -> "Flwdir":
    idxs_ds = df[ds_col].values
    idxs = df.index.values
    return Flwdir(idxs_ds=get_loc_idx(idxs=idxs, idxs_ds=idxs_ds))
```

### Current behaviour

Traceback:
```python
_____________________________________________________________________________________________________ test_model_build[piave] _____________________________________________________________________________________________________

tmpdir = local('C:\\Users\\veenstra\\AppData\\Local\\Temp\\pytest-of-veenstra\\pytest-1063\\test_model_build_piave_0'), modelname = 'piave'

    @pytest.mark.timeout(300)  # max 5 min
    @pytest.mark.slow
    @pytest.mark.parametrize("modelname", list(_models.keys()))
    def test_model_build(tmpdir, modelname):
        model_dict = _models[modelname]
    
        # Build method options
        config = model_dict["ini"]
        _, global_sect, steps = read_workflow_yaml(config)
        crs = global_sect['crs']
        network_snap_offset = global_sect['network_snap_offset']
        openwater_computation_node_distance = global_sect['openwater_computation_node_distance']
    
        # test build method
        # compare results with model from examples folder
        root = join(tmpdir, f"dflowfm_{modelname}")
        logger = logging.getLogger("hydromt")
        logger.setLevel(logging.DEBUG)
        mod1 = DFlowFMModel(
            root=root,
            mode="w",
            data_libs=[model_dict["data"]],
            network_snap_offset=network_snap_offset,
            crs=crs,
            openwater_computation_node_distance=openwater_computation_node_distance,
        )
        # Build model (now excludes global section because of pop)
>       mod1.build(steps=steps)

tests\test_hydromt.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.pixi\envs\pypi-311\Lib\site-packages\hydromt\model\model.py:323: in build
    method(**merged)
hydromt_delft3dfm\dflowfm.py:592: in setup_rivers_from_dem
    gdf_riv, _ = workflows.get_river_bathymetry(
hydromt_delft3dfm\workflows\dem.py:249: in get_river_bathymetry
    flw = pyflwdir.from_dataframe(gdf_stream.set_index("idx"))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\pyflwdir\flwdir.py:54: in from_dataframe
    return Flwdir(idxs_ds=get_loc_idx(idxs=idxs, idxs_ds=idxs_ds))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:443: in _compile_for_args
    raise e
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:376: in _compile_for_args
    return_val = self.compile(tuple(argtypes))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:908: in compile
    cres = self._compiler.compile(args, return_type)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:80: in compile
    status, retval = self._compile_cached(args, return_type)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:94: in _compile_cached
    retval = self._compile_core(args, return_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\dispatcher.py:107: in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler.py:739: in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler.py:439: in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler.py:505: in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler.py:481: in _compile_core
    raise e
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler.py:473: in _compile_core
    pm.run(self.state)
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler_machinery.py:363: in run
    raise e
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler_machinery.py:356: in run
    self._runPass(idx, pass_inst, state)
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler_lock.py:35: in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler_machinery.py:311: in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\compiler_machinery.py:272: in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
.pixi\envs\pypi-311\Lib\site-packages\numba\core\typed_passes.py:114: in run_pass
    typemap, return_type, calltypes, errs = type_inference_stage(
.pixi\envs\pypi-311\Lib\site-packages\numba\core\typed_passes.py:80: in type_inference_stage
    with callstack_ctx, warnings:
.pixi\envs\pypi-311\Lib\site-packages\numba\core\errors.py:530: in __exit__
    self.flush()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <numba.core.errors.WarningsFixer object at 0x000001ED474CC150>

    def flush(self):
        """
        Emit all stored warnings.
        """
        def key(arg):
            # It is possible through codegen to create entirely identical
            # warnings, this leads to comparing types when sorting which breaks
            # on Python 3. Key as str() and if the worse happens then `id`
            # creates some uniqueness
            return str(arg) + str(id(arg))
    
        for (filename, lineno, category), messages in sorted(
                self._warnings.items(), key=key):
            for msg in sorted(messages):
>               warnings.warn_explicit(msg, category, filename, lineno)
E               numba.core.errors.NumbaTypeSafetyWarning: unsafe cast from int32 to undefined. Precision may be lost.

.pixi\envs\pypi-311\Lib\site-packages\numba\core\errors.py:523: NumbaTypeSafetyWarning
```

### Desired behaviour

No warning. This requires pyflwdir to be fixed: https://github.com/Deltares/pyflwdir/issues/89

### Additional context

Note that this only happens in the pixi locked testing environments, not in the pixi default environment, so for instance run:
```
pixi run -e pypi-311 --locked test tests/test_hydromt.py::test_model_build[piave]
pixi run -e pypi-311 --locked test tests/test_dflowfm.py::test_setup_rivers_from_dem
```

This warning was silenced in https://github.com/Deltares/hydromt_delft3dfm/issues/103.

--- a/pixi.lock
+++ b/pixi.lock
@@ -4712,7 +4712,7 @@ packages:
 - pypi: ./
   name: hydromt-delft3dfm
   version: 0.3.1.dev0
-  sha256: 773f0570d5a6927ac977d618d2d73fb28e60649ecd9e1fdf4b1e3ba69fc838c4
+  sha256: 0dca2504078cfe1d87f09cd5d2b9a33e47c1c184267278e42e3afb2a57505e22
   requires_dist:
   - hydromt>=1.3.1
   - geopandas>=0.10,!=1.0.0

--- a/pixi.lock
+++ b/pixi.lock
@@ -4712,7 +4712,7 @@ packages:
 - pypi: ./
   name: hydromt-delft3dfm
   version: 0.3.1.dev0
-  sha256: 26131a6f73335e501233dc62432250ebc24835fa8850361649986c3f390bab8b
+  sha256: 773f0570d5a6927ac977d618d2d73fb28e60649ecd9e1fdf4b1e3ba69fc838c4
   requires_dist:
   - hydromt>=1.3.1
   - geopandas>=0.10,!=1.0.0

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -130,6 +130,7 @@ filterwarnings = [
     # note: on GitHub the warning is with int64 for `test_model_build[piave]`
     "ignore: unsafe cast from int32 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
     "ignore: unsafe cast from int64 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
+    "ignore:.*:numba.core.errors.NumbaTypeSafetyWarning", # for github windows runners
 	]
 markers = [
 	"slow: slow tests",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -109,9 +109,7 @@ exclude = ["docs", "examples", "envs", "tests", ".github"]
 [tool.pytest.ini_options]
 addopts = "--durations=0"
 filterwarnings = [
-    # TODO: cannot raise warnings as errors since NumbaTypeSafetyWarning is not properly filtered on windows github runner
-    #  to be resolved in https://github.com/Deltares/hydromt_delft3dfm/issues/289
-#	"error",
+	"error",
 	# temporarily ignore numpy incompatibility, more info in https://github.com/Deltares/dfm_tools/issues/1243
 	"ignore:numpy.ndarray size changed:RuntimeWarning",
     # filter until fix of https://github.com/Deltares/HYDROLIB-core/issues/1098
@@ -130,8 +128,8 @@ filterwarnings = [
     # ignore NumbaTypeSafetyWarning originating from pyflwdir, to be fixed in https://github.com/Deltares/hydromt_delft3dfm/issues/289
     # note: the warning starts with a space somehow
     # note: on GitHub the warning is with int64 for `test_model_build[piave]`
-#    "ignore: unsafe cast from int32 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
-#    "ignore: unsafe cast from int64 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
+    "ignore: unsafe cast from int32 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
+    "ignore: unsafe cast from int64 to undefined:numba.core.errors.NumbaTypeSafetyWarning",
 	]
 markers = [
 	"slow: slow tests",

